x/gamm/pool-models/stableswap/amm.go

```diff
@@ -0,0 +1,73 @@
+package stableswap
+
+import sdk "github.com/cosmos/cosmos-sdk/types"
+
+var cubeRootTwo, _ = sdk.NewDec(2).ApproxRoot(3)
+
+// solidly CFMM is xy(x^2 + y^2) = k
+// So we want to solve for a given addition of `b` units of y into the pool,
+// how many units `a` of x do we get out.
+// So we solve the following expression for `a`
+// xy(x^2 + y^2) = (x - a)(y + b)((x - a)^2 + (y + b)^2)
+// use the following wolfram alpha link
+// https://www.wolframalpha.com/input?i=solve+for+a%2C+xy%28x%5E2+%2B+y%5E2%29+%3D+%28x+-+a%29%28y+%2B+b%29%28%28x+-+a%29%5E2+%2B+%28y+%2Bb%29%5E2%29+
+// This returns:
+// copied from wolfram:
+// assuming (correctly) that b + y!=0
+// a = (-27 b^2 x^3 y - 27 b^2 x y^3 + sqrt((-27 b^2 x^3 y - 27 b^2 x y^3 - 54 b x^3 y^2 - 54 b x y^4 - 27 x^3 y^3 - 27 x y^5)^2 + 4 (3 b^4 + 12 b^3 y + 18 b^2 y^2 + 12 b y^3 + 3 y^4)^3) - 54 b x^3 y^2 - 54 b x y^4 - 27 x^3 y^3 - 27 x y^5)^(1/3)/(3 2^(1/3) (b + y)) - (2^(1/3) (3 b^4 + 12 b^3 y + 18 b^2 y^2 + 12 b y^3 + 3 y^4))/(3 (b + y) (-27 b^2 x^3 y - 27 b^2 x y^3 + sqrt((-27 b^2 x^3 y - 27 b^2 x y^3 - 54 b x^3 y^2 - 54 b x y^4 - 27 x^3 y^3 - 27 x y^5)^2 + 4 (3 b^4 + 12 b^3 y + 18 b^2 y^2 + 12 b y^3 + 3 y^4)^3) - 54 b x^3 y^2 - 54 b x y^4 - 27 x^3 y^3 - 27 x y^5)^(1/3)) + (b x + x y)/(b + y) and b + y!=0
+// Dev separating out terms:
+// I added {} myself, making better distinctions between entirely distinct terms. (parenthesis don't solve this, because divisions)
+// a = {(-27 b^2 x^3 y - 27 b^2 x y^3
+//			+ sqrt(
+//				(-27 b^2 x^3 y - 27 b^2 x y^3 - 54 b x^3 y^2 - 54 b x y^4 - 27 x^3 y^3 - 27 x y^5)^2
+//				+ 4 (3 b^4 + 12 b^3 y + 18 b^2 y^2 + 12 b y^3 + 3 y^4)^3)
+// 			- 54 b x^3 y^2 - 54 b x y^4 - 27 x^3 y^3 - 27 x y^5)^(1/3)
+// 		  / (3 2^(1/3) (b + y))}
+//		- {(2^(1/3) (3 b^4 + 12 b^3 y + 18 b^2 y^2 + 12 b y^3 + 3 y^4))
+//		  /(3 (b + y)
+// 			(-27 b^2 x^3 y - 27 b^2 x y^3
+// 				+ sqrt(
+//					(-27 b^2 x^3 y - 27 b^2 x y^3 - 54 b x^3 y^2 - 54 b x y^4 - 27 x^3 y^3 - 27 x y^5)^2
+//					+ 4 (3 b^4 + 12 b^3 y + 18 b^2 y^2 + 12 b y^3 + 3 y^4)^3)
+// 				- 54 b x^3 y^2 - 54 b x y^4 - 27 x^3 y^3 - 27 x y^5)^(1/3))}
+//      + {(b x + x y)/(b + y)}
+// Then notice that the two sqrt terms and surrounding items in the expression are the same.
+// So we replace them with a single term:
+// discriminant = sqrt(
```

Comment:
I don't think this is actually the discriminant, in particular because `a` isn't actually a polynomial in `x` and `y`. Maybe I should just rename this to radical